cmd/signature-v4.go

```diff
@@ -236,76 +236,40 @@ func doesPresignedSignatureMatch(hashedPayload string, r *http.Request, region s
 	t := pSignValues.Date
 	expireSeconds := int(pSignValues.Expires / time.Second)
 
-	// Construct new query.
+	// Construct the query map
 	query := make(url.Values)
-	clntHashedPayload := req.URL.Query().Get(xhttp.AmzContentSha256)
-	if clntHashedPayload != "" {
-		query.Set(xhttp.AmzContentSha256, hashedPayload)
-	}
-
-	token := req.URL.Query().Get(xhttp.AmzSecurityToken)
-	if token != "" {
-		query.Set(xhttp.AmzSecurityToken, cred.SessionToken)
-	}
-
-	query.Set(xhttp.AmzAlgorithm, signV4Algorithm)
-
-	// Construct the query.
-	query.Set(xhttp.AmzDate, t.Format(iso8601Format))
-	query.Set(xhttp.AmzExpires, strconv.Itoa(expireSeconds))
-	query.Set(xhttp.AmzSignedHeaders, getSignedHeaders(extractedSignedHeaders))
-	query.Set(xhttp.AmzCredential, cred.AccessKey+SlashSeparator+pSignValues.Credential.getScope())
-
-	// Save other headers available in the request parameters.
-	for k, v := range req.URL.Query() {
-		key := strings.ToLower(k)
-
-		// Handle the metadata in presigned put query string
-		if strings.Contains(key, "x-amz-meta-") {
-			query.Set(k, v[0])
-			continue
-		}
-
-		if strings.Contains(key, "x-amz-server-side-") {
-			query.Set(k, v[0])
-			continue
-		}
-
-		if strings.HasPrefix(key, "x-amz") {
-			continue
+	for key, val := range req.URL.Query() {
+		switch key {
+		case xhttp.AmzContentSha256:
+			{
+				if val != nil {
+					query.Set(xhttp.AmzContentSha256, hashedPayload)
+				}
+			}
+		case xhttp.AmzSecurityToken:
+			{
+				if val != nil {
+					query.Set(xhttp.AmzSecurityToken, cred.SessionToken)
+				}
+			}
+		case xhttp.AmzExpires:
+			{
+				if val != nil {
+					query.Set(xhttp.AmzExpires, strconv.Itoa(expireSeconds))
+				}
+			}
+		case xhttp.AmzSignature:
+		default:
+			{
+				query.Set(key, val[0])
+			}
 		}
-		query[k] = v
 	}
 
 	// Get the encoded query.
 	encodedQuery := query.Encode()
 
-	// Verify if date query is same.
-	if req.URL.Query().Get(xhttp.AmzDate) != query.Get(xhttp.AmzDate) {
-		return ErrSignatureDoesNotMatch
-	}
-	// Verify if expires query is same.
-	if req.URL.Query().Get(xhttp.AmzExpires) != query.Get(xhttp.AmzExpires) {
-		return ErrSignatureDoesNotMatch
-	}
-	// Verify if signed headers query is same.
-	if req.URL.Query().Get(xhttp.AmzSignedHeaders) != query.Get(xhttp.AmzSignedHeaders) {
-		return ErrSignatureDoesNotMatch
-	}
-	// Verify if credential query is same.
-	if req.URL.Query().Get(xhttp.AmzCredential) != query.Get(xhttp.AmzCredential) {
-		return ErrSignatureDoesNotMatch
-	}
-	// Verify if sha256 payload query is same.
-	if clntHashedPayload != "" && clntHashedPayload != query.Get(xhttp.AmzContentSha256) {
-		return ErrContentSHA256Mismatch
-	}
-	// Verify if security token is correct.
-	if token != "" && subtle.ConstantTimeCompare([]byte(token), []byte(cred.SessionToken)) != 1 {
-		return ErrInvalidToken
-	}
-
```

Comment:
The reason this logic exists for security reasons @ebozduman we cannot blindly assume what was client sent is correct, we need to verify the correct values.

You have also removed the code which was necessary for server to generate the query independently beyond what client actually sent. We should have the spec right to match what was sent not trust the client by default. 

So this fix is not correct.

Reply:
OK. I'll put it back.